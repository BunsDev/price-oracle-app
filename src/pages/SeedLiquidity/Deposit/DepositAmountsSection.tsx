import { useEffect, useState } from 'react';
import { BigNumber, constants, utils } from 'ethers';
import { isUndefined } from 'lodash';
import styled from 'styled-components';
import { useAccount } from 'wagmi';

import { MOBILE_MAX_WIDTH, SPACING_16, SPACING_32, SPACING_8, Typography } from '~/components/shared';
import InputNumber from '~/components/shared/InputNumber';
import { getConfig } from '~/config';
import { useContracts } from '~/hooks';
import { FeeTier, Token, UniswapPool } from '~/types';
import { sanitizeDecimals } from '~/utils';
import Balance from './Balance';
import Deposit from './Deposit';
import SubmitFormSection from './SubmitFormSection';

const Container = styled.section`
  display: grid;
  grid-template-columns: 1fr 1fr;
  column-gap: ${SPACING_16};
  row-gap: ${SPACING_8};
  padding-bottom: ${SPACING_32};

  @media (max-width: ${MOBILE_MAX_WIDTH}px) {
    display: flex;
    flex-direction: column;
  }
`;

const Title = styled(Typography).attrs({
  variant: 'x-large',
})`
  grid-column-start: span 2;
  margin-bottom: ${SPACING_16};
  text-align: start;
  margin-top: ${SPACING_16};
`;

const BalanceContainer = styled(Typography).attrs({
  variant: 'small',
  color: 'primary',
})`
  display: table;
  text-align: start;
  margin-top: ${SPACING_8};
`;

interface DepositAmountsProps {
  selectedToken: Token | undefined;
  startingPrice: BigNumber;
  uniswapPoolsForFeeTier: { [feeTier: string]: UniswapPool } | undefined;
  selectedFee: FeeTier;
}

const DepositAmountsSection = ({
  selectedToken,
  startingPrice,
  uniswapPoolsForFeeTier,
  selectedFee,
}: DepositAmountsProps) => {
  const { address: userAddress } = useAccount();
  const { erc20Service } = useContracts();
  const eth_symbol = 'WETH';
  const {
    ADDRESSES: { WETH_ADDRESS },
  } = getConfig();
  const [tokenBalance, setTokenBalance] = useState<BigNumber | undefined>(undefined);
  const [wethBalance, setWethBalance] = useState<BigNumber | undefined>(undefined);

  const onWethAmountChanged = (amount: string) => {
    const wethAmount = utils.parseEther(sanitizeDecimals(amount));
    const tokenAmount = wethAmount.mul(startingPrice).div(constants.WeiPerEther);
    if (tokenAmount.isZero()) {
      tokenInput.reset();
    } else {
      tokenInput.set(
        sanitizeDecimals(utils.formatUnits(tokenAmount, selectedToken?.decimals), selectedToken?.decimals)
      );
    }
  };

  const onTokenAmountChanged = (amount: string) => {
    const tokenAmount = utils.parseUnits(sanitizeDecimals(amount, selectedToken?.decimals), selectedToken?.decimals);
    const wethAmount = constants.WeiPerEther.mul(tokenAmount).div(startingPrice);
    if (wethAmount.isZero() || startingPrice.isZero()) {
      wethInput.reset();
    } else {
      wethInput.set(utils.formatEther(wethAmount));
    }
  };

  const tokenInput = InputNumber.useProps({
    onChange: onTokenAmountChanged,
    initialValue: '0',
  });

  const wethInput = InputNumber.useProps({
    onChange: onWethAmountChanged,
    initialValue: '0',
  });

  useEffect(() => {
    if (userAddress) {
      selectedToken &&
        erc20Service.fetchTokenBalance(selectedToken.address, userAddress).then((balance) => setTokenBalance(balance));
    }
  }, [selectedToken, tokenBalance?.toString(), userAddress]);

  useEffect(() => {
    if (userAddress) {
      erc20Service.fetchTokenBalance(WETH_ADDRESS, userAddress).then((wethBalance) => setWethBalance(wethBalance));
    }
  }, [wethBalance?.toString(), userAddress]);

  useEffect(() => {
    // Reset to undefined to restart the useEffect hook
    setTokenBalance(undefined);
  }, [selectedToken]);

  useEffect(() => {
    onTokenAmountChanged(tokenInput.value);
  }, [startingPrice]);

  const inputMaxWethBalance = () => {
    if (wethBalance) {
      const newValue = utils.formatEther(wethBalance);
      wethInput.set(newValue);
      onWethAmountChanged(newValue);
    }
  };

  const inputMaxTokenBalance = () => {
    if (tokenBalance) {
      const newValue = utils.formatUnits(tokenBalance, selectedToken?.decimals);
      tokenInput.set(newValue);
      onTokenAmountChanged(newValue);
    }
  };

  return (
    <>
      <Container>
        <Title weight='semibold'>Deposit amounts</Title>

        <div>
          <Deposit>
            <Deposit.Token src={selectedToken?.logoURI} />
            <Deposit.Amount {...tokenInput} />
            <Deposit.Symbol>
              <Typography>{selectedToken?.symbol || ''}</Typography>
            </Deposit.Symbol>
          </Deposit>
          <BalanceContainer>
            <Balance
              totalAmount={tokenBalance}
              symbol={selectedToken?.symbol || ''}
              onClick={inputMaxTokenBalance}
              decimals={selectedToken?.decimals}
            />
          </BalanceContainer>
        </div>

        <div>
          <Deposit>
            <Deposit.Token isPrice />
            <Deposit.Amount {...wethInput} />
            <Deposit.Symbol>{eth_symbol}</Deposit.Symbol>
          </Deposit>
          <BalanceContainer>
            <Balance totalAmount={wethBalance} symbol={eth_symbol} onClick={inputMaxWethBalance} />
          </BalanceContainer>
        </div>
      </Container>
      <SubmitFormSection
        tokenAmount={utils.parseUnits(
          sanitizeDecimals(tokenInput.value, selectedToken?.decimals),
          selectedToken?.decimals
        )}
        wethAmount={utils.parseEther(sanitizeDecimals(wethInput.value))}
        wethBalance={wethBalance || constants.Zero}
        tokenBalance={tokenBalance || constants.Zero}
        selectedToken={selectedToken}
        startingPrice={startingPrice}
        uniswapPoolsForFeeTier={uniswapPoolsForFeeTier}
        selectedFee={selectedFee}
      />
    </>
  );
};

export default DepositAmountsSection;
