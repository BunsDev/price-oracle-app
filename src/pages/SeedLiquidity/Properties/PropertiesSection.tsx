import { useEffect } from 'react';
import { BigNumber, utils } from 'ethers';
import { isUndefined } from 'lodash';

import {
  Container,
  NotCreatedContainer,
  SDropdownModal,
  SInputNumber,
  SPrice,
  Suffix,
} from './PropertiesSection.styles';
import FeeCard from './FeeCard';
import PropertyCard, { LockIcon, FeePropertyCard } from './PropertyCard';
import { Icon, InputNumber, Loading } from '~/components/shared';
import Dropdown from '~/components/shared/Dropdown';
import { getConfig } from '~/config';
import { FeeTier, OraclesCreated, Token, UniswapPool } from '~/types';
import { sqrtPriceX96ToPrice } from '~/utils';
import { useContracts } from '~/hooks';

interface PropertiesSectionProps {
  selectedToken: Token | undefined;
  startingPrice: BigNumber | undefined;
  setStartingPrice: (newPrice: BigNumber | undefined) => void;
  uniswapPoolsForFeeTier: { [feeTier: string]: UniswapPool } | undefined;
  setUniswapPoolsForFeeTier: (pools: { [feeTier: string]: UniswapPool } | undefined) => void;
  selectedFee: FeeTier;
  setSelectedFee: (fee: FeeTier) => void;
  oraclesCreated: OraclesCreated;
}

function PropertiesSection({
  selectedToken,
  startingPrice,
  setStartingPrice,
  uniswapPoolsForFeeTier,
  setUniswapPoolsForFeeTier,
  selectedFee,
  setSelectedFee,
  oraclesCreated,
}: PropertiesSectionProps) {
  const { uniswapService } = useContracts();
  const dropdownProps = Dropdown.useProps();

  const FEE_TIERS = getConfig().FEE_TIERS;
  const feeTierList = Object.values(FEE_TIERS);

  const isLoading = !uniswapPoolsForFeeTier;

  const setNewFee = (fee: FeeTier) => {
    dropdownProps.setShow(false);
    setSelectedFee(fee);
  };

  const onPriceChange = (newPrice: string | undefined) => {
    if (newPrice) {
      setStartingPrice(utils.parseUnits(newPrice, selectedToken?.decimals));
    } else {
      setStartingPrice(undefined);
    }
  };

  const startingPriceInput = InputNumber.useProps({ initialValue: startingPrice?.toString(), onChange: onPriceChange });

  useEffect(() => {
    if (selectedToken) {
      setUniswapPoolsForFeeTier(undefined);
      startingPriceInput.reset();
      uniswapService
        .fetchUniswapPools(selectedToken.address)
        .then((poolListMap) => setUniswapPoolsForFeeTier(poolListMap));
    }
  }, [selectedToken]);

  useEffect(() => {
    const newPrice = getPriceOfSelectedToken();
    if (newPrice) {
      startingPriceInput.set(newPrice);
      onPriceChange(newPrice);
    } else {
      startingPriceInput.set('');
      onPriceChange(undefined);
    }
  }, [selectedFee, uniswapPoolsForFeeTier]);

  const isPoolCreated = (fee: number) => uniswapPoolsForFeeTier && !isUndefined(uniswapPoolsForFeeTier[fee]);
  const selectedFeeTierExists = isPoolCreated(selectedFee.fee);

  const getPriceOfSelectedToken = () => {
    if (uniswapPoolsForFeeTier && selectedToken) {
      const uniPool = uniswapPoolsForFeeTier[selectedFee.fee];
      if (uniPool) {
        const priceInWei = sqrtPriceX96ToPrice(uniPool.pricing, uniPool.isWethToken0);
        return utils.formatUnits(priceInWei, selectedToken.decimals);
      }
    }
  };

  return (
    <Container>
      <PropertyCard data-testid='starting-price'>
        <PropertyCard.Title>Starting price</PropertyCard.Title>

        <PropertyCard.Value>
          {isLoading && <Loading />}
          {!isLoading && (
            <SInputNumber {...startingPriceInput} disabled={selectedFeeTierExists} aria-label='starting price' />
          )}
        </PropertyCard.Value>
        {selectedFeeTierExists && <LockIcon name='lock' />}
      </PropertyCard>

      <PropertyCard data-testid='token-rate'>
        <PropertyCard.Title>Rate</PropertyCard.Title>
        <PropertyCard.Value>
          1 <Suffix> ETH</Suffix>
          <Icon name='arrow-down' /* color={disabled} */ rotate={270} />
          {isLoading && <Loading />}
          {!isLoading && <SPrice>{startingPriceInput.value}</SPrice>}
          <Suffix>{selectedToken?.symbol || ''}</Suffix>
        </PropertyCard.Value>
      </PropertyCard>

      <FeePropertyCard data-testid='feeTier-selected'>
        <PropertyCard.Title>Fee Tier</PropertyCard.Title>

        <PropertyCard.Value>
          {isLoading && <Loading />}
          {!isLoading && (
            <Dropdown {...dropdownProps}>
              <Dropdown.Button>{selectedFee?.label}</Dropdown.Button>
              <SDropdownModal>
                {feeTierList.map((fee) => (
                  <FeeCard onClick={() => setNewFee(fee)} key={fee.fee}>
                    <FeeCard.FeePercentage>{fee.label}</FeeCard.FeePercentage>
                    <NotCreatedContainer>
                      {oraclesCreated[fee.fee] || <FeeCard.UsagePercentage>not created</FeeCard.UsagePercentage>}
                    </NotCreatedContainer>
                    <FeeCard.Hint>{fee.hint}</FeeCard.Hint>
                  </FeeCard>
                ))}
              </SDropdownModal>
            </Dropdown>
          )}
        </PropertyCard.Value>
        {!isLoading && <PropertyCard.Chip>{selectedFee.hint}</PropertyCard.Chip>}
      </FeePropertyCard>
    </Container>
  );
}

export default PropertiesSection;
