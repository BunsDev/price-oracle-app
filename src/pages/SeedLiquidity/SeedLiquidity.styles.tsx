import styled from 'styled-components';

import { Card, SPACING_512, SPACING_832, SPACING_24 } from '~/components/shared';

export const Container = styled(Card)`
  display: flex;
  flex-direction: column;
  justify-content: space-between;
  margin: 0 auto;
  width: 100%;
  padding: ${SPACING_24};
  max-width: ${SPACING_832};
  height: ${SPACING_512};
`;
