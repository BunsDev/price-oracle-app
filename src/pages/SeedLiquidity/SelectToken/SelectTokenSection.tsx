import styled from 'styled-components';

import {
  EthIcon,
  FONT_SIZE_24,
  Icon,
  SPACING_16,
  SPACING_8,
  TokenIcon,
  Typography,
  MOBILE_MAX_WIDTH,
} from '~/components/shared';
import Dropdown from '~/components/shared/Dropdown';
import { AlertsActions, useAppDispatch } from '~/store';
import { Token } from '~/types/Token';
import TokenList from './TokenList';

const Container = styled.section`
  grid-area: select-token;
  grid-row-gap: ${SPACING_16};
  display: flex;
  justify-content: space-between;
  font-family: ${FONT_SIZE_24};

  @media (max-width: ${MOBILE_MAX_WIDTH}px) {
    flex-direction: column;
    align-items: flex-start;
  }
`;

const Price = styled(Typography)`
  display: flex;
  align-items: center;
  column-gap: 0.5rem;
  margin-left: 0.5rem;
`;

const RightSide = styled.div`
  display: flex;
  align-items: center;
`;

const Divider = styled(Typography)`
  margin: 0 ${SPACING_8};
`;

const STokenIcon = styled(TokenIcon)`
  margin-right: ${SPACING_8};
`;

const STokenList = styled(TokenList)`
  max-height: 36rem;
`;

const PaddedTypography = styled(Typography)`
  padding: 0.1rem 0;
`;

const SDropdownModal = styled(Dropdown.Modal)`
  @media (max-width: ${MOBILE_MAX_WIDTH}px) {
    margin-top: 0;
    margin-left: 2.5rem;
  }
`;

const TokenDropdown = styled(Dropdown.Button)`
  padding-left: 0.7rem;
`;

interface SelectTokenProps {
  selectedToken: Token | undefined;
  setSelectedToken: (token: Token) => void;
}

const SelectTokenSection = ({ selectedToken, setSelectedToken }: SelectTokenProps) => {
  const dropdownProps = Dropdown.useProps();
  const dispatch = useAppDispatch();

  const onTokenSelect = (token: Token) => {
    if (token.symbol === 'WETH') {
      dispatch(AlertsActions.openAlert({ message: 'Invalid Token', type: 'error' }));
    } else {
      setSelectedToken(token);
      dropdownProps.setShow(false);
    }
  };

  return (
    <Container>
      <Typography variant='x-large' weight='semibold'>
        {'Select Token '}
        <Icon name='arrow-down' rotate={270} />
      </Typography>

      <RightSide>
        <Dropdown {...dropdownProps}>
          <TokenDropdown>
            <STokenIcon src={selectedToken?.logoURI || ''} size={FONT_SIZE_24()} />
            <PaddedTypography variant='x-large' weight='semibold'>
              {selectedToken?.symbol}
            </PaddedTypography>
          </TokenDropdown>

          <SDropdownModal>
            <Dropdown.Modal>
              <STokenList onSelect={(token: Token) => onTokenSelect(token)} />
            </Dropdown.Modal>
          </SDropdownModal>
        </Dropdown>

        <Divider variant='x-large'>{'/'}</Divider>

        <Price variant='x-large' weight='semibold'>
          <EthIcon /> WETH
        </Price>
      </RightSide>
    </Container>
  );
};

export default SelectTokenSection;
