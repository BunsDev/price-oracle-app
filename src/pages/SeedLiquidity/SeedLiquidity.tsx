import { useState } from 'react';
import { useNetwork } from 'wagmi';
import styled from 'styled-components';

import { Token } from '~/types';
import { getTokenList } from '~/utils/tokenList';
import { SPACING_24 } from '~/components/shared';
import { Container } from './SeedLiquidity.styles';
import SelectTokenSection from './SelectToken/SelectTokenSection';
import PropertiesSection from './Properties/PropertiesSection';
import DepositAmountsSection from './Deposit/DepositAmountsSection';
import SubmitFormSection from './Deposit/SubmitFormSection';
import PoolList from './SeededList/SeededList';

const SeedPage = styled.div`
  background: ${(props) => props.theme.background};
  padding: 0rem 0.8rem ${SPACING_24};
`;

function SeedLiquidity() {
  const { chain } = useNetwork();
  const defaultToken = getTokenList(chain?.id)[0];
  const [selectedToken, setSelectedToken] = useState<Token | undefined>(defaultToken);

  return (
    <SeedPage>
      <Container>
        <SelectTokenSection selectedToken={selectedToken} setSelectedToken={setSelectedToken} />
        <PropertiesSection />
        <DepositAmountsSection selectedToken={selectedToken} />
        <SubmitFormSection />
      </Container>
      <PoolList />
    </SeedPage>
  );
}

export default SeedLiquidity;
