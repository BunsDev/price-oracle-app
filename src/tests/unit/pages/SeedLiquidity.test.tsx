import { vi } from 'vitest';
import { constants } from 'ethers';

import { render, screen, within } from '~/tests';
import { feeTier, token, uniswapPoolForFeeTierMock } from '~/tests/unit/__mocks__';
import SeedLiquidity from '~/pages/SeedLiquidity';
import PropertiesSection from '~/pages/SeedLiquidity/Properties/PropertiesSection';
import DepositAmountsSection from '~/pages/SeedLiquidity/Deposit/DepositAmountsSection';
import SelectTokenSection from '~/pages/SeedLiquidity/SelectToken/SelectTokenSection';

describe('Seed Liquidity Page', () => {
  it('Should render Liquidity Page', () => {
    const { asFragment, getByText } = render(<SeedLiquidity />);

    expect(getByText(/create oracle/i)).toBeInTheDocument();
    expect(asFragment()).toMatchSnapshot();
  });
});

describe('Testing SeedLiquidity page initial state', () => {
  it('"Create Oracle" button should be disabled', () => {
    const seedLiquidity = render(<SeedLiquidity />);
    expect(seedLiquidity.getByText(/create oracle/i)).toBeInTheDocument();
    expect(seedLiquidity.getByText(/create oracle/i)).toBeDisabled();
  });

  it('Starting price, Rate and Fee Tier should start loading', () => {
    const seedLiquidity = render(<SeedLiquidity />);
    const loaderIcon = seedLiquidity.getAllByTestId('loader-icon');
    expect(loaderIcon.length).toEqual(3);
  });
});

describe('Testing <PropertiesSection />', () => {
  const functionMock = vi.fn();

  beforeEach(() => {
    render(
      <PropertiesSection
        selectedToken={token}
        startingPrice={constants.Zero}
        setStartingPrice={functionMock}
        uniswapPoolsForFeeTier={uniswapPoolForFeeTierMock}
        setUniswapPoolsForFeeTier={functionMock}
        selectedFee={feeTier}
        setSelectedFee={functionMock}
        oraclesCreated={{ [feeTier.fee]: false }}
      />
    );
  });

  it('shows the correct "Starting Price" when uniswapPool is passed', () => {
    const startingPriceSection = within(screen.getByTestId('starting-price'));
    expect(startingPriceSection.getByDisplayValue(/1094.873103/)).toBeInTheDocument();
  });

  it('shows the correct "Token Rate" when uniswapPool is passed', () => {
    const tokenRateSection = within(screen.getByTestId('token-rate'));
    expect(tokenRateSection.getByText(token.symbol)).toBeInTheDocument();
    expect(tokenRateSection.getByText(/1094.873103/)).toBeInTheDocument();
  });

  it('shows the "Fee tier" selected', () => {
    const feeTierSection = within(screen.getByTestId('feeTier-selected'));
    expect(feeTierSection.getByText(feeTier.label)).toBeInTheDocument();
    expect(feeTierSection.getByText(feeTier.hint)).toBeInTheDocument();
  });
});

describe('Testing <DepositAmountsSection />', () => {
  beforeEach(() => {
    render(
      <DepositAmountsSection
        selectedToken={token}
        startingPrice={constants.WeiPerEther}
        uniswapPoolsForFeeTier={uniswapPoolForFeeTierMock}
        selectedFee={feeTier}
        poolManagerAddress={''}
      />
    );
  });

  it('both inputs start with null value', () => {
    expect(screen.getByLabelText('token amount')).toHaveValue('');
    expect(screen.getByLabelText('weth amount')).toHaveValue('');
  });

  it('shows the selected token symbol and image', () => {
    expect(screen.getByText(token.symbol)).toBeInTheDocument();
    expect(screen.getByLabelText('token amount').previousSibling).toHaveAttribute('src', token.logoURI);
  });
});

describe('Testing <SelectTokenSection />', () => {
  const functionMock = vi.fn();

  it('shows the selected token symbol and image', () => {
    render(<SelectTokenSection selectedToken={token} setSelectedToken={functionMock} />);
    expect(screen.getByText(token.symbol)).toBeInTheDocument();
    expect(screen.getByText(token.symbol).previousSibling).toHaveAttribute('src', token.logoURI);
  });
});
