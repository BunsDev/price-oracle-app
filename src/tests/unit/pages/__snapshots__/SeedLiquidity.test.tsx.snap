// Vitest Snapshot v1

exports[`Seed Liquidity Page > Should render Liquidity Page 1`] = `
<DocumentFragment>
  <div
    class="sc-bQmOxr lmoJcE"
  >
    <div
      class="sc-iBYQkv sc-jrcTuL blHKRc epIsAC sc-ipEyDJ sc-cUEOzv hhOCqg coitij"
      color="rgba(255,255,255, 0.87)"
    >
      <div
        class="sc-fWQKxP jToRhd"
      >
        <section
          class="sc-hTBuwn gtBEoM"
        >
          <div
            class="sc-iBYQkv sc-ftTHYK cWFtuj iZMgjB"
            color="rgba(255,255,255, 0.87)"
          >
            Select Token 
            <i
              class="sc-hBxehG botUXV picon-arrow-down"
              name="arrow-down"
              rotate="270"
            />
          </div>
          <div
            class="sc-gScZFl iKDxRj"
          >
            <section
              class="sc-fnGiBr cQqyKK"
            >
              <button
                class="sc-bcXHqe sc-gswNZR sc-dkrFOg ehpQtM XXtZw iigscx sc-hiDMwi dPIZoZ"
              >
                <div
                  class="sc-bjfHbI sc-iOeugr jWtslV huDETY"
                  src="https://assets.coingecko.com/coins/images/13469/thumb/1inch-token.png?1608803028"
                />
                <div
                  class="sc-iBYQkv sc-ftTHYK cWFtuj iZMgjB sc-eJDSGI jIACl"
                  color="rgba(255,255,255, 0.87)"
                >
                  1INCH
                </div>
                <i
                  class="sc-hBxehG dBvyuW picon-chevron-down"
                  name="chevron-down"
                  rotate="0"
                />
              </button>
            </section>
            <div
              class="sc-iBYQkv sc-ftTHYK blHKRc iZMgjB sc-lbVpMG kegqDg"
              color="rgba(255,255,255, 0.87)"
            >
              /
            </div>
            <div
              class="sc-iBYQkv sc-ftTHYK cWFtuj iZMgjB sc-jfTVlA eDLA-DW"
              color="rgba(255,255,255, 0.87)"
            >
              <div
                class="sc-bjfHbI kfgBnd"
                src="https://raw.githubusercontent.com/trustwallet/assets/master/blockchains/ethereum/assets/0xC02aaA39b223FE8D0A0e5C4F27eAD9083C756Cc2/logo.png"
              />
               WETH
            </div>
          </div>
        </section>
      </div>
      <div
        class="sc-fWQKxP jToRhd"
      >
        <section
          class="sc-ehvNnt eqWQly"
        >
          <section
            class="sc-ZqFbI dxgRUR"
            data-testid="starting-price"
          >
            <div
              class="sc-iBYQkv sc-jrcTuL blHKRc epIsAC sc-hlLBRy gGoHHO"
              color="rgba(255,255,255, 0.87)"
            >
              Starting price
            </div>
            <div
              class="sc-iBYQkv sc-pyfCe blHKRc hMiFgy sc-eKJbhj gJMEzq"
              color="rgba(255,255,255, 0.87)"
            >
              <div
                class="sc-fLcnxK"
              >
                <i
                  class="sc-hBxehG sc-bBABsx ccFtCM jFMzlb picon-loader"
                  data-testid="loader-icon"
                  name="loader"
                />
              </div>
            </div>
          </section>
          <section
            class="sc-ZqFbI dxgRUR"
            data-testid="token-rate"
          >
            <div
              class="sc-iBYQkv sc-jrcTuL blHKRc epIsAC sc-hlLBRy gGoHHO"
              color="rgba(255,255,255, 0.87)"
            >
              Rate
            </div>
            <div
              class="sc-iBYQkv sc-pyfCe blHKRc hMiFgy sc-eKJbhj gJMEzq"
              color="rgba(255,255,255, 0.87)"
            >
              1 
              <div
                class="sc-iBYQkv sc-kDvujY gRZfHa faMXpT sc-gGvHcT jazvgA"
                color="rgba(255, 255, 255, 0.5)"
              >
                 ETH
              </div>
              <i
                class="sc-hBxehG botUXV picon-arrow-down"
                name="arrow-down"
                rotate="270"
              />
              <div
                class="sc-fLcnxK"
              >
                <i
                  class="sc-hBxehG sc-bBABsx ccFtCM jFMzlb picon-loader"
                  data-testid="loader-icon"
                  name="loader"
                />
              </div>
              <div
                class="sc-iBYQkv sc-kDvujY gRZfHa faMXpT sc-gGvHcT jazvgA"
                color="rgba(255, 255, 255, 0.5)"
              >
                1INCH
              </div>
            </div>
          </section>
          <section
            class="sc-ZqFbI sc-jNJNQp dxgRUR gpqJGB"
            data-testid="feeTier-selected"
          >
            <div
              class="sc-iBYQkv sc-jrcTuL blHKRc epIsAC sc-hlLBRy gGoHHO"
              color="rgba(255,255,255, 0.87)"
            >
              Fee Tier
            </div>
            <div
              class="sc-iBYQkv sc-pyfCe blHKRc hMiFgy sc-eKJbhj gJMEzq"
              color="rgba(255,255,255, 0.87)"
            >
              <div
                class="sc-fLcnxK"
              >
                <i
                  class="sc-hBxehG sc-bBABsx ccFtCM jFMzlb picon-loader"
                  data-testid="loader-icon"
                  name="loader"
                />
              </div>
            </div>
          </section>
        </section>
      </div>
      <section
        class="sc-dEVLtI fUtRQd"
      >
        <div
          class="sc-iBYQkv sc-ftTHYK cWFtuj iZMgjB sc-brePNt enZZLm"
          color="rgba(255,255,255, 0.87)"
        >
          Deposit amounts
        </div>
        <div>
          <section
            class="sc-cZFQFd gsEASS"
          >
            <div
              class="sc-bjfHbI sc-jTjUTQ jWtslV fsaznq"
              src="https://assets.coingecko.com/coins/images/13469/thumb/1inch-token.png?1608803028"
            />
            <input
              aria-label="token amount"
              class="sc-hHTYSt sc-dmctIk sc-kgTSHT sc-jRwbcX fvoNMM jgGokt isyBJj"
              placeholder="0"
              value=""
            />
            <div
              class="sc-iBYQkv sc-jrcTuL blHKRc epIsAC sc-fvEvSO fVTCGI"
              color="rgba(255,255,255, 0.87)"
            >
              <div
                class="sc-iBYQkv sc-jrcTuL blHKRc epIsAC"
                color="rgba(255,255,255, 0.87)"
              >
                1INCH
              </div>
            </div>
          </section>
          <div
            class="sc-iBYQkv sc-kDvujY blHKRc faMXpT sc-gJqSRm eigerd"
            color="rgba(255,255,255, 0.87)"
          >
            <div
              class="sc-iBYQkv sc-kDvujY cWFtuj faMXpT sc-cwSeag sc-dsHJmm myFBD fVhOTF"
              color="rgba(255,255,255, 0.87)"
            >
              Balance: - 1INCH
            </div>
          </div>
        </div>
        <div>
          <section
            class="sc-cZFQFd gsEASS"
          >
            <div
              class="sc-bjfHbI kfgBnd"
              src="https://raw.githubusercontent.com/trustwallet/assets/master/blockchains/ethereum/assets/0xC02aaA39b223FE8D0A0e5C4F27eAD9083C756Cc2/logo.png"
            />
            <input
              aria-label="weth amount"
              class="sc-hHTYSt sc-dmctIk sc-kgTSHT sc-jRwbcX fvoNMM jgGokt isyBJj"
              placeholder="0"
              value=""
            />
            <div
              class="sc-iBYQkv sc-jrcTuL blHKRc epIsAC sc-fvEvSO fVTCGI"
              color="rgba(255,255,255, 0.87)"
            >
              WETH
            </div>
          </section>
          <div
            class="sc-iBYQkv sc-kDvujY blHKRc faMXpT sc-gJqSRm eigerd"
            color="rgba(255,255,255, 0.87)"
          >
            <div
              class="sc-iBYQkv sc-kDvujY cWFtuj faMXpT sc-cwSeag sc-dsHJmm myFBD fVhOTF"
              color="rgba(255,255,255, 0.87)"
            >
              Balance: - WETH
            </div>
          </div>
        </div>
      </section>
      <section
        class="sc-bWOGAC cOjtLy"
      >
        <div
          class="sc-fcCDlF hLKyMm"
        >
          <div
            class="sc-GKYbw cmskty"
          >
            <div
              class="sc-jSUZER dotgdq"
            >
              <button
                class="sc-bcXHqe sc-gswNZR sc-dkrFOg ehpQtM XXtZw iigscx sc-gKPRtg hQQRME sc-csDkEv fQatMh"
                disabled=""
              >
                Create Oracle
              </button>
            </div>
          </div>
        </div>
      </section>
      <div
        class="sc-iBYQkv sc-kDvujY gRZfHa faMXpT sc-eGugkK bQcaFc"
        color="rgba(255, 255, 255, 0.5)"
      >
        Minimum WETH amount to create a new oracle: 25 WETH
      </div>
    </div>
  </div>
</DocumentFragment>
`;
